R/is_reproducible.R

```diff
@@ -0,0 +1,28 @@
+#' Reproducibility Check for `@code` Slot in `teal_data`
+#' @param teal_data `teal_data` object
+#' @examples
+#' tdata1 <- teal_data()
+#' tdata1 <- within(tdata1, {
+#'   a <- 1
+#'   b <- a^5
+#'   c <- list(x = 2)
+#' })
+#' is_reproducible(tdata1)
+#' @export
+is_reproducible <- function(teal_data) {
```

Comment:
Horrible name, requires care when typing.

`validate` is not good either because of conflict with `shiny`. Although maybe S4 can handle that.

Oh, one other thing. This has to be an S4 generic with method for `teal_data` and for `qenv.error`.

Also, please use `object` for the argument that accepts `teal_data` to keep in line with `qenv` methods (e.g. `eval_code)`.

Reply:
Steal from `teal.data::datanames`.

Reply:
> Also, please use `object` for the argument that accepts `teal_data` to keep in line with `qenv` methods (e.g. `eval_code)`.

On second thought, this is not `teal.code` so maybe this doesn't have to be respected. It's not in `datanames`.

Reply:
> Horrible name, requires care when typing.

These days anybody barely type the whole name, you rather start with `is_re` and then you get a recommendation by RStudio/IDEs that allows you to pick `is_reporoducible` from the list. So possibility to make a typo is invalid. Especially with a spellcheck that often highlights improperly written words. There is also GitHub pilot that will fill out `is_repro` for you, based on the content of the package files that you use. 

`is_reproducible` is a very informative name, as we frankly just check the reproducibility of the code.

Reply:
> `is_reproducible` is a very informative name, as we frankly just check the reproducibility of the code.

I disagree. _What_ "is reproducible"? You check object validity, you set the `@valid` slot.

Reply:
So maybe we should change `valid` slot name to `reproducible`? You typically check the `validation of` something. `valid`/`validation` alone does not give any information. You need to extend it by a `noun`, like `validation of reproducibility` `validation of a class name`. When it comes to `reproducibility` you know the exact meaning of this word without any extra nouns. Reproducibility is the validation check for the code to determine if it gives the expected assumed results.

Reply:
> So maybe we should change `valid` slot name to `reproducible`? 

This was considered and rejected.

> You typically check the `validation of` something. 

You check _validity_ by performing _validation_.

> When it comes to `reproducibility` you know the exact meaning of this word without any extra nouns.

Not any more than with "validation".

> Reproducibility is the validation check for the code to determine if it gives the expected assumed results.

Reproducibility is the ability of the code to reproduce the results kept in `@env`.

Reply:
I actually disagree with below

![image](https://github.com/insightsengineering/teal.data/assets/133694481/ed816813-902c-4c93-b5dc-c29fcae42a08)

validation is **the action of checking or proving the validity or accuracy of something.**

And then you need to have a custom definition of `validity` to perform `validation` of this `something`. So when you use `validate/valid` you need to add definition of `validity` in this context and the noun. For example, there is different validation of dog's health and different validation of a correct operation of the car engine. 

`validation of the object` does not mean anything. Does it mean we are checking it's class, we check whether it contains specific fields, we check whether it has extra metadata? Validation is imho fluid enough that it means nothing without the definition of validation.

My opinion is that `valid/validation` is unclear and ambiguous and `reproducibility` (which is tied to `code` and `environment`) is clear and popular enough to give a clear picture in this situation. That is why I decided to name function `is_reproducible` and would opt to rename the field from `valid` to `reproducible`.

We might not land on the same page in here, so I guess we can ask the team to express thoughts on the naming as I don't recall a discussion on below. No hard feelings at all
![image](https://github.com/insightsengineering/teal.data/assets/133694481/24b91057-3640-433b-860f-993649a7fa0d)

Reply:
> validation is **the action of checking or proving the validity or accuracy of something.**

This is exactly what I said: 
> You check validity by performing validation.

What is _validity_, i.e. what state is and what state is not valid, differs depending on context.

----

> My opinion is that `valid/validation` is unclear and ambiguous and `reproducibility` (which is tied to `code` and `environment`) is clear and popular enough to give a clear picture in this situation.

Quite the opposite. As you yourself stated above, _validation_ is checking _validity_, that is perfectly clear. And you yourself admit that "reproducibility" is to be taken in the context of code and environment, so it is _not_ a common meaning.

----

As for naming the slot "reproducible", the code may very well be reproducible and the slot can still be `FALSE` because it hasn't been challenged. Conversely, an object becomes "valid" when it passes validation.

Reply:
Yeah so `validity` is unclear, and always requires context or the definition. You would never know what `valid()` function does without inspecting it's body or documentation. However `reproducibility` executed on an object with `code` and `environment` already has some pre-assumed meaning. I'm still in favour of `is_reproducible` for the function name.

For the slot name, I agree that `reproducible` is not enough for the name. Maybe the proper and meanigfull name should be `reproducibility_check_performed`. However this is long, and the shortness of the name is a consensus of meaning and usability (the shorter the better). If not `reproducible` and maybe not `valid` for the slot name, what about `checked`, `executed`, `code_checked`, `code_executed` for the slot name?

Reply:
What about `@verified` for the slot name and `is_verified` or `verify()`/`verify_noun()` for the function name? It's really close to `@valid` and `validate()` however : P. I would love to get feedback/opinion from the bigger group. Any objections if we create a small poll?

Reply:
Much better.

Poll away.

Reply:
Hey, so far `@verified` and `verify()` are having the highest amount of votes, even though we could get 2 more votes. Will make changes to reflect those proposed names.

Reply:
301a86e4ba8c856ccd4b9cdee16fa92f21777d16